### ISSUE TITLE
Latest bash has access issues

### ISSUE THREAD

**Issue by ubitux** [Homebrew/homebrew-core#64480]
# Bug report

- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [x] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

```
ux@satan ~ % brew update
Already up-to-date.
ux@satan ~ % brew doctor
Your system is ready to brew.
ux@satan ~ % brew gist-logs bash
Error: No logs.
ux@satan ~ % brew config
HOMEBREW_VERSION: 2.5.9
ORIGIN: https://github.com/Homebrew/brew
HEAD: 2ed5d80ff2f8025823de9b9faacfca27fa006c80
Last commit: 20 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 2cd7462a0f74533e452d55bd8988590794f2c22c
Core tap last commit: 57 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit haswell
Clang: 12.0 build 1200
Git: 2.24.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
Java: 1.8.0_212
macOS: 10.15.7-x86_64
CLT: 12.1.0.0.1.1602137645
Xcode: 12.1
ux@satan ~ %
```

## What you were trying to do (and why)

Trying to change directory (from Python) before executing `/usr/local/bin/bash` fails.

```sh
mkdir -p test
python -c 'import os,subprocess;os.chdir("test");print(subprocess.check_output(["/usr/local/bin/bash", "-c", "pwd"], text=True))'
```

This worked with the previous version of `/usr/local/bin/bash`, and currently works with `/bin/bash`.

EDIT: simpler repro case: `PWD= /usr/local/bin/bash -c pwd`

## What happened (include command output)

<details>
  <summary>Command output</summary>
  <pre>
shell-init: error retrieving current directory: getcwd: cannot access parent directories: Bad file descriptor
pwd: error retrieving current directory: getcwd: cannot access parent directories: Undefined error: 0
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/Cellar/python@3.9/3.9.0_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 420, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/local/Cellar/python@3.9/3.9.0_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/usr/local/bin/bash', '-c', 'pwd']' returned non-zero exit status 1.
  </pre>
</details>

## What you expected to happen

The same as with previous bash revision from 2 days ago, or with `/bin/bash`:

```
ux@satan ~ % python -c 'import os,subprocess;os.chdir("test");print(subprocess.check_output(["/bin/bash", "-c", "pwd"], text=True))'
/Users/ux/test
```

## Step-by-step reproduction instructions (by running `brew install` commands)

```sh
brew install bash
cd
mkdir -p test
python -c 'import os,subprocess;os.chdir("test");print(subprocess.check_output(["/usr/local/bin/bash", "-c", "pwd"], text=True))'
```

**Comment by SMillerDev** [comment]
We don't do code support. If you can find an actual issue with the install we can look at it, but closing this until we have a test case that doesn't depend on the way python handles files.

**Comment by ubitux** [comment]
You don't need Python. It can be reproduced like this: `(cd $HOME/test && PWD=$HOME /usr/local/bin/bash -c pwd)`

It's a regression, it was working in the previous revision.

You can also try with the bash system which works too: `(cd $HOME/test && PWD=$HOME /bin/bash -c pwd)`

**Comment by gromgit** [comment]
Can't reproduce:
```
$ brew ls --versions bash
bash 5.0.18
$ mkdir -p ~/test
$ (cd $HOME/test && PWD=$HOME /usr/local/bin/bash -c pwd)
/Users/aho/test
```

**Comment by ubitux** [comment]
> Can't reproduce:
> 
> ```
> $ brew ls --versions bash
> bash 5.0.18
> $ mkdir -p ~/test
> $ (cd $HOME/test && PWD=$HOME /usr/local/bin/bash -c pwd)
> /Users/aho/test
> ```

Did you `brew upgrade bash`? This is a regression from about 2 days ago

**Comment by gromgit** [comment]
Ah.
```sh
$ brew reinstall bash
==> Downloading https://homebrew.bintray.com/bottles/bash-5.0.18.catalina.bottle.1.tar.gz
==> Downloading from https://d29vzk4ow07wi7.cloudfront.net/ca367f96ad5c40ae70b155af67d0a4ab772b16c55ccfc99b9317ae0afbbbc493?response-content-disposition=attachment%3Bfilename%3D%22bash-5.0.18.catalina.bottle.1.tar.gz%22&Policy=eyJTdGF0ZW1lbnQiOiBbeyJSZXNvdXJjZSI6Imh0dHAqOi8
######################################################################## 100.0%
==> Reinstalling bash 
==> Pouring bash-5.0.18.catalina.bottle.1.tar.gz
🍺  /usr/local/Cellar/bash/5.0.18: 150 files, 10.4MB

# After opening new shell session...
$ brew ls --version bash
bash 5.0.18
$ (cd $HOME/test && PWD=$HOME /usr/local/bin/bash -c pwd)
shell-init: error retrieving current directory: getcwd: cannot access parent directories: Bad file descriptor
pwd: error retrieving current directory: getcwd: cannot access parent directories: Undefined error: 0
```
That's disturbing.

**Comment by ubitux** [comment]
> That's disturbing.

BTW, we can also do `unset PWD` instead of setting it to an invalid value.

Can we re-open the issue?

**Comment by fxcoudert** [comment]
The issue is probably not linked with https://github.com/Homebrew/homebrew-core/pull/64345, which fixes the detection of `strsignal` and `snprintf` with Xcode 12.

However, it's possible that `configure` is misdetecting some other functions. The regression is likely linked to Xcode 12, as the previous build was with Xcode 11.

See for example https://github.com/NixOS/nixpkgs/commit/dff0ba38a243603534c99f2cf1af81c793cad98e and https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/merge_requests/443

**Comment by fxcoudert** [comment]
Proposed fix: https://github.com/Homebrew/homebrew-core/pull/64549

**Comment by jonchang** [comment]
Thanks for the fix FX.

**Comment by fxcoudert** [comment]
If you're impacted by this issue: `brew reinstall bash` will fix it

**Comment by matchbookmac** [comment]
Working for me now as well, thank you.

**Comment by ubitux** [comment]
> If you're impacted by this issue: `brew reinstall bash` will fix it

Is there no package release mechanism in brew so that packaging fixes/changes would end up being pulled automatically by a brew update/upgrade? (basically `5.0.18-1` → `5.0.18-2`)

**Comment by fxcoudert** [comment]
There is, but it would update the vast majority of users who have absolutely nothing to gain. See https://github.com/Homebrew/homebrew-core/issues/64582#issuecomment-725540911

**Comment by znep** [comment]
@fxcoudert I know there are trade-offs here around forcing an update that is unnecessary for most people due to the short time window where it was broken, as for a data point (and to maybe help people with similar symptoms to mine find this) it took me most of a day to even trace the problem down to being in bash. 

The symptoms I saw were infinite loops in calling `dirname` from the [asdf](https://github.com/asdf-vm/asdf) package manager, but only when compiling a particular Rust package ([rustler](https://github.com/rusterlium/rustler)).  And a builtin in bash being broken is way down there on the list of suspects.

Once I identified it was in bash, it was pretty easy to work around by compiling it myself but it took further digging to figure out there was a new build in homebrew that wasn't applied when I tried an upgrade.  I'm not an expert with homebrew so didn't expect to find a new build with the same version, even though that is obviously an established pattern.

Thanks for the fix.
